### PR TITLE
Support for legendre (mpz_legendre)

### PR DESCRIPTION
**Pull request by ruescasd**

### CODE REVIEW

**Comment by dragonsinth** [comment]
Mostly LG.  Any reason not to do jacobi while we're at it?

**Comment by ruescasd** [comment]
+jacobi and kronecker, tested on libgmp 5.1.3

File: jnagmp/src/main/java/com/squareup/jnagmp/LibGmp.java [hunk 1]
```diff
@@ -324,8 +324,19 @@ public static void __gmpz_export(Pointer rop, Pointer countp, int order, int siz
...
-   * for such p it’s identical to the Jacobi symbol.
+   * Calculate the Jacobi symbol (a/b). This is defined only for b odd.
+   */
+  public static native int __gmpz_jacobi(mpz_t a, mpz_t p);
+
+  /**
+   * Calculate the Jacobi symbol (a/b) with the Kronecker extension (a/2)=(2/a)
+   * when a odd, or (a/2)=0 when a even.
+   */
+  public static native int __gmpz_kronecker_si(mpz_t a, long p);
```
Comment (dragonsinth) [inline jnagmp/src/main/java/com/squareup/jnagmp/LibGmp.java]: There's a mismatch between the doc and param names here, "b" vs. "p".  I notice even the GMP docs are slightly inconsistent, using P for legendre and B for jacobi and kronecker.  Maybe we should follow the wikipedia nomenclature and use P for legendre and jacobi and N for kronecker?

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 2]
```diff
@@ -148,13 +150,45 @@ public static BigInteger modInverse(BigInteger val, BigInteger modulus) {
...
+  }
+
+  /**
+   * Calculate kronecker symbol a|p.
+   *
+   * @param a an integer
+   * @param p the modulus
+   * @return a|p
+   */
+  public static int kronecker(BigInteger a, long p) {
```
Comment (dragonsinth) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: I think we should just use mpz_kronecker (const mpz_t A, const mpz_t B) to allow both arguments to be BigInteger.  Do you have an compelling reason otherwise?  In generally we've opted for this approach instead of trying to target precision-specific functions.

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 3]
```diff
@@ -141,6 +150,59 @@ public static BigInteger modInverse(BigInteger val, BigInteger modulus) {
...
+    return INSTANCE.get().kroneckerImpl(a, p);
+  }
+
+  /**
+   * Calculate legendre symbol a|p.
```
Comment (dragonsinth) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: it looks like internally, GMP defines legendre and jacobi to be the exact same implementation.  That makes me wonder if we shouldn't expose jacobi only; otherwise you might expect legendre to fail on non-prime P and be surprised when it doesn't fail.

File: jnagmp/src/test/java/com/squareup/jnagmp/GmpTest.java [hunk 4]
```diff
@@ -173,6 +176,87 @@ public void testModInverseArithmeticException() {
...
+
+    assertEquals(0, kronecker(BigInteger.valueOf(0), -8));
+    assertEquals(1, kronecker(BigInteger.valueOf(1), -8));
+    assertEquals(0, kronecker(BigInteger.valueOf(2), -8));
+    assertEquals(-1, kronecker(BigInteger.valueOf(3), -8));
+    assertEquals(0, kronecker(BigInteger.valueOf(4), -8));
+    assertEquals(-1, kronecker(BigInteger.valueOf(5), -8));
+    assertEquals(0, kronecker(BigInteger.valueOf(6), -8));
+    assertEquals(1, kronecker(BigInteger.valueOf(7), -8));
+    assertEquals(0, kronecker(BigInteger.valueOf(8), -8));
```
Comment (dragonsinth) [inline jnagmp/src/test/java/com/squareup/jnagmp/GmpTest.java]: Can you add a test for a negative A, and also a P value of 0?

**Comment by dragonsinth** [comment]
@JakeWharton @swankjesse this is getting close, do you have a CLA for this?

**Comment by ruescasd** [comment]
> I think we should just use mpz_kronecker (const mpz_t A, const mpz_t B) to allow both arguments to be BigInteger. Do you have an compelling reason otherwise? In generally we've opted for this approach instead of trying to target precision-specific functions.

The reason is that the mpz_kronecker (const mpz_t A, const mpz_t B) function does not appear to be in the library, listing symbols only this appears:

249: 0000000000019ce0   357 FUNC    GLOBAL DEFAULT   11 __gmpz_ui_kronecker
   483: 0000000000019e50   327 FUNC    GLOBAL DEFAULT   11 __gmpz_kronecker_si
   558: 0000000000019fa0   276 FUNC    GLOBAL DEFAULT   11 __gmpz_kronecker_ui
   631: 0000000000019b20   444 FUNC    GLOBAL DEFAULT   11 __gmpz_si_kronecker

**Comment by ruescasd** [comment]
> it looks like internally, GMP defines legendre and jacobi to be the exact same implementation. That makes me wonder if we shouldn't expose jacobi only; otherwise you might expect legendre to fail on non-prime P and be surprised when it doesn't fail.

Agreed.

**Comment by dragonsinth** [comment]
Oh hey, it looks like:

```
#define mpz_kronecker mpz_jacobi  /* alias */
```

So perhaps the trick is, only expose mpz_jacobi in LibGmp, and provide both jacobi and kronecker at the Gmp API layer, the only difference being that jacobi enforces argument restrictions.

**Comment by ruescasd** [comment]
Yes, here's why it has this name:

This code does triple duty as mpz_jacobi, mpz_legendre and mpz_kronecker. For ABI compatibility, the link symbol is __gmpz_jacobi, not __gmpz_kronecker, even though the latter would be more logical.

**Comment by ruescasd** [comment]
I'm leaving out kronecker as the implementation in mpz_jacobi is wrong, failing on this test:
`
`assertEquals(-1, kronecker(BigInteger.valueOf(3), BigInteger.valueOf(-8)));``

mpz_jacobi returns 1, whereas mpz_kronecker_si correctly returns -1.

**Comment by ruescasd** [comment]
> EDIT: I tried this out, and it worked right up until:

    assertEquals(-1, kronecker(BigInteger.valueOf(3), BigInteger.valueOf(-8)));

I'd like to understand more about why this failed.

Hadn't seen this update, I ran into the same problem. As I mentioned earlier, the implementation in jacobi is wrong, but mpz_kronecker_si is correct.

File: jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java [hunk 5]
```diff
@@ -141,6 +148,27 @@ public static BigInteger modInverse(BigInteger val, BigInteger modulus) {
   }
 
   /**
+   * Calculate jacobi symbol a|p.
+   *
+   * @param a must be positive
+   * @param p the modulus must be odd
+   * @return a|p
+   * @throws IllegalArgumentException if a is not positive, or p is not positive, or p is not odd
```
Comment (dragonsinth) [inline jnagmp/src/main/java/com/squareup/jnagmp/Gmp.java]: `if a is negative, p is negative, or p is not odd`

**Comment by dragonsinth** [comment]
One tiny doc nit (which I can fix), otherwise LGTM.
I just need to get confirmation that your CLA was signed, and I'll merge.
Thanks a bunch for the contribution!

**Comment by ruescasd** [comment]
For the record the problem with the kronecker assert failure is not a gmp problem. I suspect it has to do with the conversion of BigIntegers into mpz_t that has issues with negative numbers. Don't have time to look at this right now, just warning in case it comes up in the future.

**Comment by dragonsinth** [comment]
Thanks, that's super helpful to know!  I'll look into our BigInteger -> mpz_t conversion code and see if I can figure out what the problem is.  If I can fix it, exposing kronecker seems like a good thing.

**Comment by dragonsinth** [comment]
@ruescasd Fixed our handling of negative numbers, such that the kronecker tests all pass now.  We'll need to rebase this however.  Do you mind if I rebase/squash all your commits?

I think we could get away with a single kronecker function now that covers everything.

**Comment by ruescasd** [comment]
Sure go ahead

**Comment by dragonsinth** [comment]
Subsumed by https://github.com/square/jna-gmp/pull/16

**Comment by swankjesse** [comment]
CLA signed 4/25/2016 15:47:58.
